firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,471 @@ use crate::{
                 };
 
-                self.poll_next(_cx)
-            }
+                matched_key_nibbles.push(nib);
+
+                let child = merkle
+                    .get_node(child_addr)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-            IteratorState::Iterating {
-                check_child_nibble,
-                visited_node_path,
-            } => {
-                let next = find_next_result(merkle, visited_node_path, check_child_nibble)
-                    .map_err(|e| api::Error::InternalError(Box::new(e)))
-                    .transpose();
+                let partial_key = match child.inner() {
+                    NodeType::Branch(branch) => &branch.path,
+                    NodeType::Leaf(leaf) => &leaf.path,
+                    NodeType::Extension(_) => {
+                        panic!("extension nodes shouldn't exist")
+                    }
+                };
 
-                Poll::Ready(next)
+                let (comparison, new_unmatched_key_nibbles) =
+                    is_prefix(partial_key.iter(), unmatched_key_nibbles);
+                unmatched_key_nibbles = new_unmatched_key_nibbles;
+
+                match comparison {
+                    Ordering::Less => {
+                        // `child` is before `key`.
+                        return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+                    }
+                    Ordering::Equal => {
+                        // `child` is a prefix of `key`.
+                        matched_key_nibbles.extend(partial_key.iter().copied());
+                        node = child;
+                        node_addr = child_addr;
+                    }
+                    Ordering::Greater => {
+                        // `child` is after `key`.
+                        let child = merkle
+                            .get_node(child_addr)
+                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+                        iter_stack.push(IterationNode::Unvisited {
+                            key: matched_key_nibbles
+                                .iter()
+                                .copied()
+                                .chain(partial_key.iter().copied())
+                                .collect(),
+                            node: child,
+                        });
+
+                        return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+                    }
+                }
             }
-        }
+            NodeType::Leaf(leaf) => {
+                let (comparison, _) = is_prefix(leaf.path.iter(), unmatched_key_nibbles);
+
+                match comparison {
+                    Ordering::Less | Ordering::Equal => {}
+                    Ordering::Greater => {
+                        let node = merkle
+                            .get_node(node_addr)
+                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
```

Comment:
🤔, am I missing something or are you redundantly populating `node`?

Reply:
removed